src/App.jsx

```diff
@@ -1,46 +1,245 @@
-  const completeTodo = (id) => {
-    const updatedTodos = todoList.map((todo) => {
+  const completeTodo = async (id) => {
+    const todoToComplete = todoList.find((todo) => todo.id === id);
+    const originalTodo = { ...todoToComplete };
+
+    const optimisticTodos = todoList.map((todo) => {
       if (todo.id === id) {
         return { ...todo, isCompleted: true };
       }
       return todo;
     });
-    setTodoList(updatedTodos);
+
+    setTodoList(optimisticTodos);
+
+    const payload = {
+      records: [
+        {
+          id: id,
+          fields: {
+            title: todoToComplete.title,
+            isCompleted: true,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: 'PATCH',
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(payload),
+    };
+
+    try {
+      const resp = await fetch(url, options);
+
+      if (!resp.ok) {
+        throw new Error(resp.statusText || 'Failed to complete todo');
+      }
+    } catch (error) {
+      console.error(error);
+      setErrorMessage(`${error.message}. Reverting todo...`);
+
+      const revertedTodos = todoList.map((todo) => {
+        if (todo.id === id) {
+          return originalTodo;
+        }
+        return todo;
+      });
+
+      setTodoList(revertedTodos);
+    }
   };
 
-  const updateTodo = (editedTodo) => {
-    const updatedTodos = todoList.map(todo => {
+  const updateTodo = async (editedTodo) => {
+    const originalTodo = todoList.find((todo) => todo.id === editedTodo.id);
+
+    const optimisticTodos = todoList.map((todo) => {
       if (todo.id === editedTodo.id) {
         return editedTodo;
       }
       return todo;
     });
-    setTodoList(updatedTodos);
+
+    setTodoList(optimisticTodos);
+
+    const payload = {
+      records: [
+        {
+          id: editedTodo.id,
+          fields: {
+            title: editedTodo.title,
+            isCompleted: editedTodo.isCompleted,
+          },
+        },
+      ],
+    };
+
+    const options = {
+      method: 'PATCH',
+      headers: {
+        Authorization: token,
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(payload),
+    };
+
+    try {
+      const resp = await fetch(url, options);
+
+      if (!resp.ok) {
+        throw new Error(resp.statusText || 'Failed to update todo');
+      }
+
+      const { records } = await resp.json();
+      const updatedTodo = {
+        id: records[0].id,
+        ...records[0].fields,
+      };
+
+      if (!updatedTodo.isCompleted) {
+        updatedTodo.isCompleted = false;
+      }
+    } catch (error) {
+      console.error(error);
+      setErrorMessage(`${error.message}. Reverting todo...`);
+
+      const revertedTodos = todoList.map((todo) => {
+        if (todo.id === originalTodo.id) {
+          return originalTodo;
+        }
+        return todo;
+      });
+
+      setTodoList(revertedTodos);
```

Comment:
Remember that with state setter functions you want to pass in a copy of the data set. Apply this to other instance in your code base where you are not passing a copy to state setter functions.